src/ext.rs

```diff
@@ -98,25 +92,73 @@ where
 impl<F> Future for Timeout<F>
 where
     F: TryFuture,
-    F::Error: From<io::Error>,
 {
-    type Output = Result<F::Ok, F::Error>;
+    type Output = Result<F::Ok, Waited<F::Error>>;
 
     fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
         match self.as_mut().future().try_poll(cx) {
             Poll::Pending => {}
-            other => return other,
+            other => return other.map_err(Waited::InnerError),
         }
 
         if self.timeout().poll(cx).is_ready() {
-            let err = Err(io::Error::new(io::ErrorKind::TimedOut, "future timed out").into());
+            let err = Err(Waited::TimedOut);
             Poll::Ready(err)
         } else {
             Poll::Pending
         }
     }
 }
 
+/// Enum returned by a future with a timeout
+#[derive(Debug)]
+pub enum Waited<E> {
```

Comment:
Calling this `TimeoutError` might perhaps be easier to follow.

Reply:
That makes sense. I'll take a look at this when I get home.
Are we comfortable extending TryFuture rather than Future, or do we need to discuss this?

Reply:
Come to think of it, I would probably consider waiting for trait specialisation before extending Future. Then impl Future<Output=x> can return Result<x, TimeoutError> **and** impl TryFuture<I, E> can flatten the output to return Result<I, TimeoutError>.